src/profilers.cpp

```diff
@@ -2,19 +2,336 @@
 // Copyright (C) CERN. See LICENSE for details.
 
 #include "profilers.hpp"
+#include "server/server.hpp"
 #include <cstdlib>
 #include <future>
 #include <iostream>
 #include <signal.h>
 #include <sys/wait.h>
 #include <fcntl.h>
 #include <nlohmann/json.hpp>
+#include <Poco/Net/StreamSocket.h>
+#include <boost/core/demangle.hpp>
+#include <boost/algorithm/string.hpp>
+#include <boost/program_options/parsers.hpp>
+
 
 #ifndef APERF_SCRIPT_PATH
 #define APERF_SCRIPT_PATH "."
 #endif
 
+
+
 namespace aperf {
+
+
+  /**
+    Constructs a MetricReader object.
+
+    @param metric_command  Metric command to be executed
+    @param metric_name     Alias for the metric command given by the user                
+    @param freq            The frequency at which the metric command will be "sampled"                   
+    @param regex           Regular expression provided by the user 
+                           to parse the output from the metric command
+    @param server_buffer   Size of the connection buffer to the server        
+  */
+
+  MetricReader::MetricReader(std::string metric_command, 
+                            std::string metric_name, 
+                            int freq,
+                            std::string regex,
+                            unsigned int server_buffer){
+    
+    this->metric_command = metric_command;
+    this->metric_name = metric_name;
+    this->freq = freq;
+    this->regex = regex;
+    this->server_buffer = server_buffer;
+    this->name = "MetricReader";
+
+    this->errorMessages= {
+      {201, "(ERROR_SDERR) stderr log file could not be created."},
+      {202, "(ERROR_STDOUT_DUP2) stdout could not be redirected to metric-reader."},
+      {203, "(ERROR_STDERR_DUP2) stderr could not be redirected to a file."},
+      {204, "(ERROR_NO_NUMBER_REGEX) metric command returned no number."},
+      {205, "(ERROR_TOO_MANY_NUMBERS_REGEX) metric command returned too many numbers."},
+      {206, "(ERROR_CONVERSION_TO_FLOAT)metric reader failed to parse float from metric command."},
+      {207, "(ERROR_PARSING_CONNECTION_INSTRS) metric reader failed to parse instructions to connect to server."},
+      {208, "(ERROR_USER_REGEX_MATCH) input regex did not match any metric command output."},
+      {209, "(ERROR_PIPE_METRIC_EXEC) Could not establish communication pipe between metric-exec "
+                "and metric-read in \"" + this->get_name() + "\"! Terminating "
+                "the profiled command wrapper."}
+    };
```

Comment:
A nice idea, I like it :)

I'd just either move the ```handle_errors``` method to a lambda function inside ```this->process``` (so that you don't have to add extra methods and fields to your class) or move the error constants (lines 122-130) to a wider scope and use them for both ```this->errorMessages``` and the code inside ```this->process```.

Reply:
Moved  handle errors inside process